// vectra/lib/WebFetcher.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebFetcher = void 0;
const axios_1 = __importDefault(require("axios"));
const cheerio = __importStar(require("cheerio"));
const turndown_1 = __importDefault(require("turndown"));
const ALLOWED_CONTENT_TYPES = [
    "text/html",
    "application/json",
    "application/xml",
    "application/javascript",
    "text/plain",
];
const DEFAULT_HEADERS = {
    Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8",
    "Accept-Encoding": "gzip, deflate, sdch, br",
    "Accept-Language": "en-US,en;q=0.8,ms;q=0.6",
    "Alt-Used": "LEAVE-THIS-KEY-SET-BY-TOOL",
    Connection: "keep-alive",
    Host: "LEAVE-THIS-KEY-SET-BY-TOOL",
    Referer: "https://www.google.com/",
    "Sec-Fetch-Dest": "document",
    "Sec-Fetch-Mode": "navigate",
    "Sec-Fetch-Site": "cross-site",
    "Upgrade-Insecure-Requests": "1",
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36",
};
class WebFetcher {
    constructor(config) {
        this._config = Object.assign({
            htmlToMarkdown: true,
            summarizeHtml: false,
        }, config);
    }
    fetch(uri, onDocument) {
        return __awaiter(this, void 0, void 0, function* () {
            const httpClient = axios_1.default.create({
                validateStatus: () => true,
            });
            // Clone headers to avoid mutating the original
            const headers = Object.assign({}, DEFAULT_HEADERS, this._config.headers);
            // get hostname from url
            const host = new URL(uri).hostname;
            headers['Host'] = host;
            headers['Alt-Used'] = host;
            // Fetch page and check for errors
            const response = yield httpClient.get(uri, Object.assign({ headers }, this._config.requestConfig));
            if (response.status >= 400) {
                throw new Error(`Site returned an HTTP status of ${response.status}`);
            }
            // Check for valid content type
            const contentType = response.headers['content-type'];
            const contentTypeArray = contentType.split(';');
            if (!contentTypeArray[0] || !ALLOWED_CONTENT_TYPES.includes(contentTypeArray[0])) {
                throw new Error(`Site returned an invalid content type of ${contentType}`);
            }
            // Convert content type to doc type
            const docType = contentTypeArray[0] != 'text/plain' ? contentTypeArray[0].split('/')[1] : undefined;
            if (docType == 'html' && this._config.htmlToMarkdown) {
                const text = this.htmlToMarkdown(response.data, uri);
                return yield onDocument(uri, text, 'md');
            }
            else {
                const text = response.data;
                return yield onDocument(uri, text, docType);
            }
        });
    }
    htmlToMarkdown(html, baseUrl) {
        var _a;
        // Parse HTML and remove scripts
        const $ = cheerio.load(html, { scriptingEnabled: true });
        // Remove scripts and convert relative links to absolute
        $('script').remove();
        $('a').each((i, elem) => {
            const $el = $(elem);
            const href = $el.attr("href");
            if (href && !href.startsWith("http")) {
                // Try converting to an absolute link
                try {
                    $el.attr("href", new URL(href, baseUrl).toString());
                }
                catch (_a) {
                    // Leave as is
                }
            }
        });
        // Convert to markdown
        const body = (_a = $('body').html()) !== null && _a !== void 0 ? _a : '';
        const turndownService = new turndown_1.default({
            hr: '\n\n---\n\n',
        });
        convertTables(turndownService);
        const md = turndownService.turndown(body);
        // Remove any overly long header text
        const contentStart = Math.min(md.indexOf('\n'), md.indexOf(' '));
        if (contentStart > 64) {
            return md.slice(contentStart);
        }
        else {
            return md;
        }
    }
}
exports.WebFetcher = WebFetcher;
function convertTables(turndownService) {
    turndownService.addRule('tableCell', {
        filter: ['th', 'td'],
        replacement: function (content, node) {
            return cell(content, node);
        }
    });
    turndownService.addRule('tableRow', {
        filter: 'tr',
        replacement: function (content, node) {
            var borderCells = '';
            var alignMap = { left: ':--', right: '--:', center: ':-:' };
            if (isHeadingRow(node)) {
                for (var i = 0; i < node.childNodes.length; i++) {
                    var border = '---';
                    var align = (node.childNodes[i].getAttribute('align') || '').toLowerCase();
                    if (align)
                        border = alignMap[align] || border;
                    borderCells += cell(border, node.childNodes[i]);
                }
            }
            return '\n' + content + (borderCells ? '\n' + borderCells : '');
        }
    });
    turndownService.addRule('table', {
        filter: ['table'],
        replacement: function (content, node) {
            // Ensure there are no blank lines
            content = content.replace('\n\n', '\n');
            return '\n\n' + content + '\n\n';
        }
    });
    turndownService.addRule('tableSection', {
        filter: ['thead', 'tbody', 'tfoot'],
        replacement: function (content) {
            return content;
        }
    });
}
const indexOf = Array.prototype.indexOf;
const every = Array.prototype.every;
// A tr is a heading row if:
// - the parent is a THEAD
// - or if its the first child of the TABLE or the first TBODY (possibly
//   following a blank THEAD)
// - and every cell is a TH
function isHeadingRow(tr) {
    var parentNode = tr.parentNode;
    return (parentNode.nodeName === 'THEAD' ||
        (parentNode.firstChild === tr &&
            (parentNode.nodeName === 'TABLE' || isFirstTbody(parentNode)) &&
            every.call(tr.childNodes, function (n) { return n.nodeName === 'TH'; })));
}
function isFirstTbody(element) {
    var previousSibling = element.previousSibling;
    return (element.nodeName === 'TBODY' && (!previousSibling ||
        (previousSibling.nodeName === 'THEAD' &&
            /^\s*$/i.test(previousSibling.textContent))));
}
function cell(content, node) {
    var index = indexOf.call(node.parentNode.childNodes, node);
    var prefix = ' ';
    if (index === 0) {
        prefix = '| ';
    }
    return cleanContent(prefix + content + ' |');
}
function cleanContent(content) {
    let output = '';
    const chars = ['\n', '\r', '\t', '\f', '\v', '\u00a0', '\u2028', '\u2029', ' '];
    for (let i = 0; i < content.length; i++) {
        if (chars.includes(content[i])) {
            if (output[output.length - 1] != ' ') {
                output += ' ';
            }
            continue;
        }
        else {
            output += content[i];
        }
    }
    return output;
}
